### ISSUE TITLE
Is it possible to cancel a running script?

### ISSUE THREAD

**Issue by KennethScott** [keithbabinec/PowerShellHostedRunspaceStarterkits#1]
Is it possible somehow to cancel a currently running script?  I was reviewing CustomRunspaceStarter/HostedRunspace.cs and my suspicion is that we could hook in with a [CancellationToken](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-5.0) on the InvokeAsync call, but I'm a bit out of my depth here.  Any info would be appreciated.

Thank you!

**Comment by keithbabinec** [comment]
Hi @KennethScott -- great question, happy to help. 

The PowerShell.InvokeAsync() method does not support a cancellation token as a parameter. However since it is an async call, it will return a Task object that is basically a handle to the async operation.

You can either:
1) Pass in a cancel token to the Task.Wait() method, and then 'trip' that cancel token via another thread (such as from a UI event or a timer outside the thread).
or
2) Pass in a duration/Timespan to the Task.Wait() method, which auto-cancels the task after the requested duration has expired.

Here is a quick example of what calling InvokeAsync() might look like with the cancel token:

```csharp
// upstream somewhere
var cancelTokenSource = new System.Threading.CancellationTokenSource();

// inside RunScript()
// NOTE: the 'await' call is gone here so we can access the task object.
Task<PSDataCollection<PSObject>> asyncTask = ps.InvokeAsync();
asyncTask.Wait(cancelTokenSource.Token); // Wait on the task until natural completion OR cancel token fires.
asyncTask.Result // this object should contain the original PSObject results (pipeline output).

// somewhere in a different thread
// invoke the cancel
cancelTokenSource.Cancel();
```

After invoking the cancel token OR the duration timeout is reached, a TaskCanceledException will be thrown so you need to handle that.

Let me know if that's enough info and gets you unblocked.

**Comment by KennethScott** [comment]
Great, thank you for the detailed explanation!  I'll give it a shot.

**Comment by KennethScott** [comment]
Hi Keith, I just wanted to say thanks for your work on this project and also the above info on canceling scripts.  I've incorporated these concepts into a new app I just published called [SpecOps](https://github.com/KennethScott/SpecOps).  It's a simple .NET web app for hosting scripts for easier access by users.  It's still a work in progress, but it is fully functional.  I credited your work in the README.  Thanks again!
